README.md
# meteor-sitemaps [![Build Status](https://api.travis-ci.org/gadicc/meteor-sitemaps.svg)](https://travis-ci.org/gadicc/meteor-sitemaps)

*Quickly create dynamic ("real-time") sitemaps using your own functions.*

Copyright (c) 2013 by Gadi Cohen <dragon@wastelands.net> and released under the MIT license (see LICENSE.txt).

### Quick Start

A simple example is shown below, with more complicated examples further down.  Sitemaps are dynamic (generated "on-the-fly" at request time), intended for moderate use with the latest data from your database.  If you have more than a few hundred pages and/or are crawled at high frequency, you may be better off creating a static sitemap.

```bash
meteor add gadicohen:sitemaps
```

1. Create `server/sitemaps.js` which contains something like:

```js
sitemaps.add('/sitemap.xml', function() {
  // required: page
  // optional: lastmod, changefreq, priority, xhtmlLinks, images, videos
  return [
    { page: '/x', lastmod: new Date() },
    { page: '/y', lastmod: new Date(), changefreq: 'monthly' },
    { page: '/z', lastmod: new Date().getTime(), changefreq: 'monthly', priority: 0.8 },
    // https://support.google.com/webmasters/answer/178636?hl=en
    { page: '/pageWithViedeoAndImages',
      images: [
        { loc: '/myImg.jpg', },        // Only loc is required
        { loc: '/myOtherImg.jpg',      // Below properties are optional
          caption: "..", geo_location: "..", title: "..", license: ".."}
      ],
      videos: [
        { loc: '/myVideo.jpg', },      // Only loc is required
        { loc: '/myOtherVideo.jpg',    // Below properties are optional
          thumbnail_loc: "..", title: "..", description: ".." etc }
      ]
    },
    // https://support.google.com/webmasters/answer/2620865?hl=en
    { page: 'lang/english', xhtmlLinks: [
      { rel: 'alternate', hreflang: 'de', href: '/lang/deutsch' },
      { rel: 'alternate', hreflang: 'de-ch', href: '/lang/schweiz-deutsch' },
      { rel: 'alternate', hreflang: 'en', href: '/lang/english' }
    ]}
  ];
});
```

You can call `sitemaps.add()` as many times as you like.
More details on the format below.
Note that the `url` is automatically added to the data served from
`/robots.txt` (since 0.0.4, using the robots.txt smart package).

**Important**: The above example uses a brand new Date() for every link.  *This
is just for demonstration purposes*.  Of course you should use the real date
of the last page update (`updatedAt` from the database?).  If you always use
the current time, Google will penalize you (or at the very least, ignore this
field on future crawls).

### Full Usage

```js
// To add a sitemap
sitemaps.add(url, list);

// To compress sitemap as gzip files. Note this will apply to all sitemap files
sitemaps.config('gzip', true/false); // default to false
```

#### URL

The obvious example is `/sitemap.xml`.  You can call the function
more than once to have many different (types of) sitemaps.  The URL is added
to the output of /robots.txt automatically (since 0.0.4).

Note that the location is
[important](http://www.sitemaps.org/protocol.html#location).  A sitemap can only
reference other URLs in its own path or descendant paths.  e.g. `/sitemap.xml`
can reference all URLs on the site.  `/articles/sitemap.xml` can only reference
other pages in the `/articles/` directory/path/route.

#### List (Array or Function)

The list can either be an array in the following format, or a function that
returns an array in the following format (e.g. a function that iterates over
information in a Collection).

```js
[
  {
    // Required.  http[s]://sitename.com automatically prepended */
    page: '/pageName',
    // Optional.  Timestamp of when the page was last modified.
    lastmod: new Date(),         // or new Date().getTime()
    // Optional.  always, hourly, daily, weekly, monthly, yearly, never
    // http://www.sitemaps.org/protocol.html#changefreqdef
    changefreq: 'monthly',
    // Optional.  http://www.sitemaps.org/protocol.html#prioritydef
    priority: 0.8
    // Optional.  https://support.google.com/webmasters/answer/2620865
    // Again, the base URL is automatically prepended to the href key
    xHtmlLinks: [
      { ref: 'alternate', 'hreflang': 'en', 'href': 'en/blah' },
      { ref: 'alternate', 'hreflang': 'de', 'href': 'de/blah' }
    ],
    // Optional.  https://support.google.com/webmasters/answer/2620865?hl=en
    // Again, the base URL is automatically prepended to the loc key
    images: [
      { loc: '/myImg.jpg' },      // Only loc is required
      { loc: '/myOtherImg.jpg',   // Below properties are optional
        caption: "..", geo_location: "..", title: "..", license: ".."}
    ],
    // Optional.  https://support.google.com/webmasters/answer/80472?hl=en
    // Again, the base URL is automatically prepended to loc, *_loc
    videos: [
      { loc: '/myVideo.jpg' },    // Only loc is required
      { loc: '/myOtherVideo.jpg'  // Below properties are optional
        thumbnail_loc: "..", title: "..", description: "..", etc: ".." }
    ]
  }
]
```

Other options might come soon, e.g. to automatically use routes in your app
to build the sitemap.

#### Example (from Meteorpedia)

```js
// To compress all sitemap as gzip file
sitemaps.config('gzip', true);

sitemaps.add('/mw_AllPages_sitemap.xml.gz', function(req) {
  // NOTE: req is the Node request object, use it when you need to retrieve information (host, port, protocol ...)
  // check out more in ./example/example.js

  var out = [], pages = WikiPages.find().fetch();
  _.each(pages, function(page) {
    out.push({
      page: 'read/' + page.name,
      lastmod: page.lastUpdated
    });
  });
  return out;
});
```

You can see this output here: http://www.meteorpedia.com/mw_AllPages_sitemap.xml

#### Locations (page, loc, href, etc)

Anywhere where a url can be provided, you can provide a "relative URL" (with
or without a leading /), and Meteor.absoluteUrl() will be prepended.  You
can override this by calling `sitemaps.config('rootUrl', 'myRootUrl')`.  For
individual links, providing an absoluet URL (beginning with `http://` or
`https://`) will avoid this behaviour.  URI components are escaped for you.

## Contributors

Thanks to @zol, @tarang, @dandv, @DirkStevens, @picsoung, @SashaG for various
PRs as listed in [History.md](History.md).


History.md
## vNEXT

## v0.0.26 (2016-10-25)

* GZIP compression support.  Big thanks to @nlhuykhang, this has been on the
  TODO list for a long time.  See the README for instructions and examples,
  but in short, `sitemaps.config('gzip', true)`.  This will probably be the
  default in the future.  (#48, #43)

* Sitemap `Content-Type` now explicitly specifies `charset=UTF-8`.

## v0.0.25 (2016-10-20)

* Pass `req` object to `.add()` callback, useful for dealing with multi-domain
  apps.  thanks @nlhuykhang (#47)

## v0.0.24 (2016-03-14)

* api.use('underscore') to work in Meteor 1.2+ and avoid conflicts with
  globally installed lodash (if it exists).

## v0.0.23 (2015-08-28)

* Release under MIT license.

## v0.0.22 (2015-Jul-10)

* Escape & in loc tag (must be &amp; in XML) - fixes #29

## v0.0.21 (2015-Apr-22)

* Use encodeURI() enstead of escape() - thanks @picsoung (#24)
* Allow custom ROOT_URLs - thanks @picsoung (#25)
* New sitemaps.config(key, value) and sitemaps.config({key1: value1, etc})
* New sitemaps.config('rootUrl', myRootUrl) to configure ROOT_URL (#26)
* Don't prepend ROOT_URL on absolute URLS (#27)
* Bump robots-txt from 0.0.8 to 0.0.9
* Add dependency on check package - thanks @SashaG (#28)
* Finally added a comprehensive test suite :)

## v0.0.20 (2014-Nov-06)

* Support for images, thanks @DirkStevens (#22)
* SUpport for videos
* Clean up indentation, refactor URL escaping, images/videos

## v0.0.19 (2014-Oct-11)

* Whitespace per Meteor standards, thanks @dandv (#15, #16)
* Use Meteor.absoluteUrl() to generate URLs (#17).

## v0.0.18 (2014-Aug-08)

* Typos, spacing, quote, etc... thanks @dandv (#12)
* Upload to Meteor 0.9.x package server, thanks @tarang (#11, #14)

## v0.0.17 (2014-Jul-17)

* Multiple protocols reported by reverse proxy (closes #10, thanks tarangp!)

## v0.0.16 (2014-Apr-08)

* Fix documentation to not require `.getTime()` on `new Date()`
* Run user code in a Fiber so Collection calls work (#5)
* Removed < 0.6.5 support :)
* What happened in 0.0.15? :)

## v0.0.14 (2014-Feb-19)

* Fix for double slash issue, thanks @zol (#3)

## v0.0.13 (2014-Jan-24)

* Support for rel="alternate" and hreflang="x" (#2)


tests.js
var parseString = Meteor.wrapAsync(Npm.require('xml2js').parseString);
var zlib = Npm.require('zlib');
var http = Npm.require('http');

// Expects a "url" relative to ROOT_URL.
// Inherently validates XML and throws an error if invalid
function fetch(url) {
  var res = HTTP.get(Meteor.absoluteUrl() + url);
  return parseString(res.content);
}

function addAndFetch(data) {
  var url = Random.id();
  sitemaps.config('gzip', false);
  sitemaps.add(url, data);
  return fetch(url);
}

function addAndFetchGzip(data, cb) {
  var route = Meteor.absoluteUrl();
  var url = Random.id();

  sitemaps.config('gzip', true);
  sitemaps.add(url, data);

  http.get(route + url, Meteor.bindEnvironment(function(res) {
    var re = [];

    res.on('data', function(chunk) {
      re.push(Buffer.from(chunk));
    });

    res.on('end', Meteor.bindEnvironment(function() {
      zlib.gunzip(Buffer.concat(re), Meteor.bindEnvironment(function(err, buff) {
        var xml = parseString(buff.toString());

        cb && cb(xml);
      }));

    }));

  }));
}


/* urls */

Tinytest.add('sitemaps - config - sitemaps.config(key, value);', function(test) {
  sitemaps.config('testkey', 123);
  test.equal(sitemaps._config.testkey, 123);
});

Tinytest.add('sitemaps - config - sitemaps.config({key1: value1, key2: value2})', function(test) {
  sitemaps.config({testkey1: 1, testkey2: 2});
  test.equal(sitemaps._config.testkey1, 1);
  test.equal(sitemaps._config.testkey2, 2);
});

Tinytest.add('sitemaps - urls - relative with leading /', function(test) {
  test.equal(sitemaps._prepareUrl('/test'),
    Meteor.absoluteUrl() + 'test');
});

Tinytest.add('sitemaps - urls - relative without leading /', function(test) {
  test.equal(sitemaps._prepareUrl('test'),
    Meteor.absoluteUrl() + 'test');
});

Tinytest.add('sitemaps - urls - uri escaped', function(test) {
  test.equal(sitemaps._prepareUrl('space space.html'),
    Meteor.absoluteUrl() + 'space%20space.html');
});

Tinytest.add('sitemaps - urls - & must be escaped', function(test) {
  test.equal(sitemaps._prepareUrl('?arg1=val1&arg2=val2&arg3=val3'),
    Meteor.absoluteUrl() + '?arg1=val1&amp;arg2=val2&amp;arg3=val3');
});

Tinytest.add('sitemaps - urls - custom root', function(test) {
  var rootUrl = 'http://custom.com/';
  sitemaps.config('rootUrl', rootUrl);
  test.equal(sitemaps._prepareUrl('custom'), rootUrl+'custom');
  sitemaps.config('rootUrl', undefined);
});

Tinytest.add('sitemaps - urls - absolute url', function(test) {
  test.equal(sitemaps._prepareUrl('http://cdn.com/test'),
    'http://cdn.com/test');
});

Tinytest.add('sitemaps - urls - used by loc/xhtml/image/video', function(test) {
  var sitemapUrl = 'usedBy.xml';
  var unescapedUrl = ' ';
  var escapedUrl = Meteor.absoluteUrl() + '%20';

  sitemaps.config('gzip', false);
  sitemaps.add(sitemapUrl, [
    {
      page: unescapedUrl,
      images: [ { loc: unescapedUrl } ],
      videos: [ { loc: unescapedUrl } ],
      xhtmlLinks: [ { href: unescapedUrl } ]
    }
  ]);

  var sitemap = fetch(sitemapUrl);
  var page = sitemap.urlset.url[0];

  test.equal(page.loc[0], escapedUrl);
  test.equal(page['xhtml:link'][0].$.href, escapedUrl);
  test.equal(page['image:image'][0]['image:loc'][0], escapedUrl);
  test.equal(page['video:video'][0]['video:loc'][0], escapedUrl);
});

/* general */

Tinytest.add('sitemaps - README example generates valid sitemap', function(test) {
  sitemaps.config('gzip', false);
  sitemaps.add('/readme.xml', function() {
    // required: page
    // optional: lastmod, changefreq, priority, xhtmlLinks, images, videos
    return [
      { page: '/x', lastmod: new Date() },
      { page: '/y', lastmod: new Date(), changefreq: 'monthly' },
      { page: '/z', lastmod: new Date().getTime(), changefreq: 'monthly', priority: 0.8 },
      // https://support.google.com/webmasters/answer/178636?hl=en
      { page: '/pageWithViedeoAndImages',
        images: [
          { loc: '/myImg.jpg', },        // Only loc is required
          { loc: '/myOtherImg.jpg',      // Below properties are optional
            caption: "..", geo_location: "..", title: "..", license: ".."}
        ],
        videos: [
          { loc: '/myVideo.jpg', },      // Only loc is required
          { loc: '/myOtherVideo.jpg',    // Below properties are optional
            thumbnail_loc: "..", title: "..", description: "..", etc: '..' }
        ]
      },
      // https://support.google.com/webmasters/answer/2620865?hl=en
      { page: 'lang/english', xhtmlLinks: [
        { rel: 'alternate', hreflang: 'de', href: '/lang/deutsch' },
        { rel: 'alternate', hreflang: 'de-ch', href: '/lang/schweiz-deutsch' },
        { rel: 'alternate', hreflang: 'en', href: '/lang/english' }
      ]}
    ];
  });

  test.ok(fetch('readme.xml'));
});

Tinytest.add('sitemaps - sitemap+page relative URL with leading /', function(test) {
  sitemaps.config('gzip', false);
  sitemaps.add('/sitemap1.xml', [ { page: '/page2' } ]);
  var sitemap = fetch('sitemap1.xml'); // throws error if /test1.xml isn't served
  var url = sitemap.urlset.url[0].loc[0];
  test.equal(url, Meteor.absoluteUrl('page2'));
});

Tinytest.add('sitemaps - sitemap+page relative URL without leading /', function(test) {
  sitemaps.config('gzip', false);
  sitemaps.add('sitemap2.xml', [ { page: 'page2' } ]);
  var sitemap = fetch('sitemap2.xml'); // throws error if /test1.xml isn't served
  var url = sitemap.urlset.url[0].loc[0];
  test.equal(url, Meteor.absoluteUrl('page2'));
});

/* namespaces */

Tinytest.add('sitemaps - namespaces - xmlns always', function(test) {
  var sitemap = addAndFetch([ { page: 'x' } ]);
  test.equal(sitemap.urlset.$.xmlns, 'http://www.sitemaps.org/schemas/sitemap/0.9');
  test.equal(Object.keys(sitemap.urlset.$).length, 1);
});

Tinytest.add('sitemaps - namespaces - xmlns:xhtml on xhtmlLinks', function(test) {
  var sitemap = addAndFetch([ { page: 'x',
    xhtmlLinks: [ { rel: 'alternate', hreflang: 'en', href: 'xen' } ] } ]);
  test.equal(sitemap.urlset.$['xmlns:xhtml'], 'http://www.w3.org/1999/xhtml');
  test.equal(Object.keys(sitemap.urlset.$).length, 2);
});

Tinytest.add('sitemaps - namespaces - xmlns:image on image', function(test) {
  var sitemap = addAndFetch([ { page: 'x', images: [ { loc: 'ximg' } ] } ]);
  test.equal(sitemap.urlset.$['xmlns:image'],
    'http://www.google.com/schemas/sitemap-image/1.1');
  test.equal(Object.keys(sitemap.urlset.$).length, 2);
});

Tinytest.add('sitemaps - namespaces - xmlns:video on video', function(test) {
  var sitemap = addAndFetch([ { page: 'x', videos: [ { loc: 'xvid' } ] } ]);
  test.equal(sitemap.urlset.$['xmlns:video'],
    'http://www.google.com/schemas/sitemap-video/1.1');
  test.equal(Object.keys(sitemap.urlset.$).length, 2);
});

/* gzip-namespaces */

Tinytest.addAsync('sitemaps-gzip - namespaces - xmlns always', function(test, next) {
  addAndFetchGzip([ { page: 'x' } ], function(xml) {
    test.equal(xml.urlset.$.xmlns, 'http://www.sitemaps.org/schemas/sitemap/0.9');
    test.equal(Object.keys(xml.urlset.$).length, 1);
    next();
  });
});

Tinytest.addAsync('sitemaps-gzip - namespaces - xmlns:xhtml on xhtmlLinks', function(test, next) {
  addAndFetchGzip([{
    page: 'x',
    xhtmlLinks: [ { rel: 'alternate', hreflang: 'en', href: 'xen' } ]
  }], function(xml) {
    test.equal(xml.urlset.$['xmlns:xhtml'], 'http://www.w3.org/1999/xhtml');
    test.equal(Object.keys(xml.urlset.$).length, 2);
    next();
  });
});

Tinytest.addAsync('sitemaps-gzip - namespaces - xmlns:image on image', function(test, next) {
  addAndFetchGzip([ { page: 'x', images: [ { loc: 'ximg' } ] } ], function(xml) {
    test.equal(xml.urlset.$['xmlns:image'],
      'http://www.google.com/schemas/sitemap-image/1.1');
    test.equal(Object.keys(xml.urlset.$).length, 2);
    next();
  });
});

Tinytest.addAsync('sitemaps-gzip - namespaces - xmlns:video on video', function(test, next) {
  addAndFetchGzip([ { page: 'x', videos: [ { loc: 'xvid' } ] } ], function(xml) {
    test.equal(xml.urlset.$['xmlns:video'],
      'http://www.google.com/schemas/sitemap-video/1.1');

    test.equal(Object.keys(xml.urlset.$).length, 2);
    next();
  });
});


sitemaps.js
/*
 * http://en.wikipedia.org/wiki/Site_map
 * http://www.sitemaps.org/index.html
 */

sitemaps = {
  _list: {},
  _config: {
    rootUrl: undefined,
    gzip: false,
  },
  _configHooks: {}
};

function configSet(key, value) {
  if (sitemaps._configHooks[key])
    sitemaps._configHooks[key](key, value, sitemaps._config[key]);
  sitemaps._config[key] = value;
}

sitemaps.config = function(key, value) {
  if (!value && _.isObject(key)) {
    for (k in key)
      configSet(k, key[k]);
  } else {
    configSet(key, value);
  }
};

if (typeof Number.lpad === "undefined") {
  Number.prototype.lpad = function(length) {
    "use strict";
    var str = this.toString();
    while (str.length < length) {
      str = "0" + str;
    }
    return str;
  };
}

var urlStart = Meteor.absoluteUrl();

sitemaps._configHooks.rootUrl = function(key, value) {
  urlStart = value || Meteor.absoluteUrl();
};

var prepareUrl = sitemaps._prepareUrl = function(url) {
  if (url.match(/^https?:\/\//))
    return url;
  else {
    return urlStart + encodeURI(url.replace(/^\//, '')).replace(/&/g, '&amp;');
  }
};

// TODO: 1) gzip, 2) sitemap index + other types + sitemap for old content
var Fiber = Npm.require('fibers');
var zlib = Npm.require('zlib');
var stream = Npm.require('stream');

var resAsXML = function resAsXML(res, data) {
  res.writeHead(200, {
    'Content-Type': 'application/xml; charset=UTF-8',
  });

  res.end(data);
  return;
};

var resAsGZip = function resAsGZip(res, data) {
  // create read stream from text
  var s = new stream.Readable();
  s.push(data);
  s.push(null);

  res.writeHead(200, {
    'Content-Type': 'application/xml; charset=UTF-8',
    'Content-Encoding': 'gzip',
  });

  // pipe to zlib to compress data
  s.pipe(zlib.createGzip()).pipe(res);
};

WebApp.connectHandlers.use(function(req, res, next) {
  new Fiber(function() {
    "use strict";
    var out, pages, urls;

    urls = _.keys(sitemaps._list);
    if (!_.contains(urls, req.url))
      return next();

    pages = sitemaps._list[req.url];
    if (_.isFunction(pages))
      pages = pages(req);
    else if (!_.isArray(pages))
      throw new TypeError("sitemaps.add() expects an array or function");

    // The header is added later once we know which namespaces we need
    out = '';
    var namespaces = {};

    var w3cDateTimeTS, date;
    _.each(pages, function(page) {

      out += '  <url>\n'
        + '    <loc>' + prepareUrl(page.page) + '</loc>\n';

      if (page.lastmod) {
        date = new Date(page.lastmod);
        w3cDateTimeTS = date.getUTCFullYear() + '-'
          + (date.getUTCMonth()+1).lpad(2) + '-'
          + date.getUTCDate().lpad(2) + 'T'
          + date.getUTCHours().lpad(2) + ':'
          + date.getUTCMinutes().lpad(2) + ':'
          + date.getUTCSeconds().lpad(2) + '+00:00';
        out += '    <lastmod>' + w3cDateTimeTS + '</lastmod>\n';
      }

      if (page.changefreq)
        out += '    <changefreq>' + page.changefreq + '</changefreq>\n';

      if (page.priority)
        out += '    <priority>' + page.priority + '</priority>\n';

      if (page.xhtmlLinks) {
        namespaces.xhtml = true;
        if (!_.isArray(page.xhtmlLinks))
          page.xhtmlLinks = [page.xhtmlLinks];
        _.each(page.xhtmlLinks, function(link) {
          out += '    <xhtml:link \n';
          if (link.href)
            link.href = prepareUrl(link.href);
          for (var key in link)
            out += '      ' + key + '="' + link[key] + '"\n';
          out += '      />\n';
        });
      }

      _.each(['image', 'video'], function(tag) {
        var tagS = tag+'s';
        if (page[tagS]) {
          namespaces[tag] = true;
          if (!_.isArray(page[tagS]))
            page[tagS] = [page[tagS]];

          _.each(page[tagS], function(data) {
            out += '      <'+tag+':'+tag+'> \n';

            for (var key in data) {
              if (key == 'loc' || key.match(/_loc$/))
                data[key] = prepareUrl(data[key]);
              out += '        <'+tag+':'+key+'>' + data[key] + '</'+tag+':'+key+'>\n';
            }

            out += '      </'+tag+':'+tag+'> \n';
          });
        }
      });

      out  += '   </url>\n\n';
    });

    out += '</urlset>\n';

    // We do this last so we know which namesapces to add
    var header = '<?xml version="1.0" encoding="UTF-8"?>\n'
      + '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"';

    if (namespaces.xhtml)
      header += '\n  xmlns:xhtml="http://www.w3.org/1999/xhtml"';
    if (namespaces.image)
      header += '\n  xmlns:image="http://www.google.com/schemas/sitemap-image/1.1"';
    if (namespaces.video)
      header += '\n  xmlns:video="http://www.google.com/schemas/sitemap-video/1.1"';
    header += '>\n';

    out = header + out;

    // response as gzip
    if (sitemaps._config.gzip) {
      resAsGZip(res, out);
      return;
    }

    resAsXML(res, out);
    return;
  }).run();
});

sitemaps.add = function(url, func) {
  "use strict";
  check(url, String);
  if (url.charAt(0) !== '/')
    url = '/' + url;

  sitemaps._list[url] = func;
  robots.addLine('Sitemap: ' + prepareUrl(url));
};

/*
sitemaps.add('/sitemap.xml', function() {
  // 'page' is reqired
  // 'lastmod', 'changefreq', 'priority' are optional.
  return [
    { page: 'x', lastmod: new Date().getTime() },
    { page: 'y', lastmod: new Date().getTime(), changefreq: 'monthly' },
    { page: 'z', lastmod: new Date().getTime(), changefreq: 'monthly', priority: 0.8 }
  ];
});
*/


example/example.js
if (Meteor.isServer) {
  var Pages = new Meteor.Collection('pages');

  if (Pages.find().count() == 0) {
    Pages.insert({ page: 'x', lastmod: new Date().getTime() });
    Pages.insert({ page: '/y', lastmod: new Date().getTime(), changefreq: 'monthly' });
    Pages.insert({ page: 'z', lastmod: new Date(), changefreq: 'monthly', priority: 0.8 });
  }

  // From README
  sitemaps.add('/sitemap.xml', function() {
    // required: page
    // optional: lastmod, changefreq, priority, xhtmlLinks, images, videos
    return [
      { page: '/x', lastmod: new Date() },
      { page: '/y', lastmod: new Date(), changefreq: 'monthly' },
      { page: '/z', lastmod: new Date().getTime(), changefreq: 'monthly', priority: 0.8 },
      // https://support.google.com/webmasters/answer/178636?hl=en
      { page: '/pageWithViedeoAndImages',
        images: [
          { loc: '/myImg.jpg', },        // Only loc is required
          { loc: '/myOtherImg.jpg',      // Below properties are optional
            caption: "..", geo_location: "..", title: "..", license: ".."}
        ],
        videos: [
          { loc: '/myVideo.jpg', },      // Only loc is required
          { loc: '/myOtherVideo.jpg',    // Below properties are optional
            thumbnail_loc: "..", title: "..", description: "..", etc: '..' }
        ]
      },
      // https://support.google.com/webmasters/answer/2620865?hl=en
      { page: 'lang/english', xhtmlLinks: [
        { rel: 'alternate', hreflang: 'de', href: '/lang/deutsch' },
        { rel: 'alternate', hreflang: 'de-ch', href: '/lang/schweiz-deutsch' },
        { rel: 'alternate', hreflang: 'en', href: '/lang/english' }
      ]}
    ];
  });

  var getFullLink = function(req) {
    if (!req) {
      return '';
    }

    var headers = req.headers;

    var host = headers && headers.host;

    var protocol = headers && headers['x-forwarded-proto'];

    if (protocol && protocol.match((/^https/))) {
      protocol = 'https://';
    } else {
      protocol = 'http://';
    }

    return protocol + host;
  };

  sitemaps.config('gzip', true);

  sitemaps.add('/sitemapDB.xml.gz', function(req) {
    // NOTE: to get the exact domain, in case you have a multiple-domain app
    var fullLink = getFullLink(req);

    var out = [];

    out.push({
      page: fullLink + '/',
    });

    var out = [], pages = Pages.find().fetch();
    _.each(pages, function(page) {
      out.push({
        page: page.page,
        lastmod: page.lastmod
      });
    });
    return out;
  });
}
